sql/analyzer/apply_join.go

```diff
@@ -245,3 +246,153 @@ func normalizeSelectSingleRel(ctx *sql.Context, a *Analyzer, n sql.Node, scope *
+// expressions before checking the normalized subquery and its hoisted
+// filters for further hoisting.
+func recurseSubqueryForOuterFilters(n sql.Node, a *Analyzer, scope *Scope) (sql.Node, transform.TreeIdentity, []sql.Expression, error) {
+	var hoistFilters []sql.Expression
+	lowestAllowedIdx := len(scope.Schema())
+	var inScope TableAliases
+	ret, same, err := transform.Node(n, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		sq, _ := n.(*plan.SubqueryAlias)
+		if sq != nil {
+			subScope := scope.newScopeFromSubqueryAlias(sq)
+			newQ, same, hoisted, err := recurseSubqueryForOuterFilters(sq.Child, a, subScope)
+			if err != nil {
+				return n, transform.SameTree, err
+			}
+			if same {
+				return n, transform.SameTree, nil
+			}
+			if len(hoisted) > 0 {
+				hoistFilters = append(hoistFilters, hoisted...)
+			}
+			ret := *sq
+			ret.Child = newQ
+			return &ret, transform.NewTree, nil
+		}
+		f, _ := n.(*plan.Filter)
+		if f == nil {
+			return n, transform.SameTree, nil
+		}
+
+		var keepFilters []sql.Expression
+		allSame := transform.SameTree
+		queue := splitConjunction(f.Expression)
+		for len(queue) > 0 {
+			e := queue[0]
+			queue = queue[1:]
+
+			var not bool
+			if n, ok := e.(*expression.Not); ok {
+				not = true
+				e = n.Child
+			}
+
+			// (1) normalize subquery expressions
+			// (1a) recurse downwards
+			// (1b) add hoisted to queue
+			// (1c) standardize subquery expression for hoisting
+			var sq *plan.Subquery
+			switch e := e.(type) {
+			case *plan.InSubquery:
+				sq, _ = e.Right.(*plan.Subquery)
+			case *plan.ExistsSubquery:
+				sq = e.Query
+			default:
+			}
+			if sq != nil {
+				children := e.Children()
+				subScope := scope.newScopeFromSubqueryExpression(n)
+				newQ, same, hoisted, err := recurseSubqueryForOuterFilters(sq.Query, a, subScope)
+				if err != nil {
+					return n, transform.SameTree, err
+				}
+				allSame = allSame && same
+				if len(hoisted) > 0 {
+					newScopeFilters, _, err := FixFieldIndexesOnExpressions(scope, a, n.Schema(), hoisted...)
+					if err != nil {
+						return n, transform.SameTree, err
+					}
+					queue = append(queue, newScopeFilters...)
+				}
+				newSq := sq.WithQuery(newQ)
+				children[len(children)-1] = newSq
+				e, _ = e.WithChildren(children...)
+			}
+
+			if not {
+				e = expression.NewNot(e)
+			}
+
+			if lowestAllowedIdx == 0 {
+				// cannot hoist filters above root scope
+				keepFilters = append(keepFilters, e)
+				continue
+			}
+
+			// (2) evaluate if expression hoistable
+			var outerRef bool
+			var innerRef bool
+			if inScope == nil {
+				var err error
+				inScope, err = getTableAliases(n, nil)
+				if err != nil {
+					return n, transform.SameTree, err
+				}
+			}
+			transform.InspectExpr(e, func(e sql.Expression) bool {
+				gf, _ := e.(*expression.GetField)
+				if gf == nil {
+					return false
+				}
+				if _, ok := inScope[strings.ToLower(gf.Table())]; ok {
+					innerRef = true
```

Comment:
So, if an expression `e` has any `GetField` expression in it, for a table that is present below the current node (which contains expression `e`), we mark that as an inner ref that is **not** eligible for hoisting? That makes sense. 

Seems like that also means this optimization won't (yet) work for something like `select * from xy where x in (select x from xy where x = 1)`? (I'm not complaining or saying we need to cover that in the first pass, just checking my understanding of the code.)

Reply:
I tried originally using getField.Index for the in/out of scope comparison, but those values are not yet reliable at this rule. For the example above, the `x=1` should still resolve to the inner table field, so I don't think this hoist rule would apply.